doc/man/kdesudo.1
'\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH kdesudo 1 2007-03-26 "" ""
.SH NAME
kdesudo \- a sudo frontend for KDE
.SH SYNOPSIS
'nh
.fi
.ad l
\fBkdesudo\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[
\fB\fIGeneric-options\fB\fR
] [
\fB\fIkdesudo-options\fB\fR
]
'in \n(.iu-\nxu
.ad b
'hy
.SH DESCRIPTION
kdesudo is a frontend for sudo for the KDE desktop.
.SH OPTIONS
.TP 
\*(T<\fB\-\-help\fR\*(T> 
Show help about options
.TP 
\*(T<\fB\-\-help\-qt\fR\*(T> 
Show Qt specific options
.TP 
\*(T<\fB\-\-help\-kde\fR\*(T> 
Show KDE specific options
.TP 
\*(T<\fB\-\-help\-all\fR\*(T> 
Show all options
.TP 
\*(T<\fB\-\-author\fR\*(T> 
Show author information
.TP 
\*(T<\fB\-v\fR\*(T>, \*(T<\fB\-\-version\fR\*(T> 
Show version information
.TP 
\*(T<\fB\-\-license\fR\*(T> 
Show license information
.TP 
\*(T<\fB\-\-\fR\*(T> 
Indicates end of options
.TP 
\*(T<\fB\-c \(Fo command \(Fc\fR\*(T> 
Specifies the command to run
.TP 
\*(T<\fB\-u \(Fo user \(Fc\fR\*(T> 
Specifies the target uid [default is root]
.TP 
\*(T<\fB\-n\fR\*(T> 
Do not keep password
.TP 
\*(T<\fB\-s\fR\*(T> 
Forgets all passwords
.TP 
\*(T<\fB\-p \(Fo priority \(Fc\fR\*(T> 
Set priority value: between 0 and 100, 0 is lowest [default is 50]
.TP 
\*(T<\fB\-\-nonewdcop\fR\*(T> 
Let command use existing dcopserver
.TP 
\*(T<\fB\-\-comment \(Fo comment \(Fc\fR\*(T> 
Comment to display in the dialog box
.TP 
\*(T<\fB\-\-noignorebutton\fR\*(T> 
Do not display \(Fo ignore \(Fc button
.TP 
\*(T<\fB\-\-attach \(Fo window_id \(Fc \fR\*(T>
Makes the dialog transient for an X app specified by winid
.TP 
\*(T<\fB\-i \(Fo icon_name \(Fc\fR\*(T> 
Specify icon to use in the password dialog
.TP 
\*(T<\fB\-d\fR\*(T> 
Do not show the command to be run in the dialog
.TP 
\*(T<\fB\-r\fR\*(T> 
Use realtime scheduling
.TP 
\*(T<\fB\-f \(Fo file \(Fc\fR\*(T> 
Use target UID if \(Fo file \(Fc is not writeable
.TP 
\*(T<\fB\-t\fR\*(T> 
Enable terminal output (no password keeping)
.PP
.TP 
\*(T<\fB\-u\fR\*(T> 
Sets a runas user
.SH COPYRIGHT
This manual page was written by Anthony Mercatante
<\*(T<tonio@ubuntu.com\*(T>> for the
Ubuntu system (but may be used by others).
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU General Public License,
Version 2 or any later version published by the Free Software Foundation.
.PP
On Debian systems, the complete text of the GNU General Public
License can be found in
\*(T<\fI/usr/share/common\-licenses/GPL\fR\*(T>.


ChangeLog
KdeSudo 3.4.2.4
------------
* Drop old and not really useful docbook documentation.
* Update french and portuguese translations
* Fix password encoding problem (LP: #400851), patch donated by Felix Geyer.

KdeSudo 3.4.2.3
------------
* Add --desktop cmd line option to manually specify the desktop file used for
    resolving name and icon

KdeSudo 3.4.2.2
------------
* Fixes LP bug #422919 set bugAddress to lp (suggested by Dario Andres)

KdeSudo 3.4.2.1
------------
* Fixes Debian #525292 for real. Avoids having the token go on clear on the
  commandline by using QProcess'es with builtin pipe.

KdeSudo 3.4.2
-------------
Fixes LP bugs #281877, #258799
Fixes LP bug #285084 (patch by Florian Reinhard)
Fixes Debian #525292 (patch by Didier 'OdyX' Raboud)

KdeSudo 3.4
-----------
* Fixed bug when bad password typed in.
  - Closes Launchpad #333203

KdeSudo 3.3.2
-----------
* Added po file for the following locales:
  - ar, cs, da, de, el, en_GB, es, et, fa, fi, fr, gl, he, hr, hu, id,
    is, it, ja, jv, kk, ko, lt, lv, ms, nb, nl, oc, pl, pt, pt_BR, ro,
    ru, sk, sv, tl, tr, uk, zh_CN, zh_TW
  - Those do result from an export of Rosetta (Launchpad) and might be
    incomplete in some case.
* Reworked the docs, and added manpage in the tarball instead of just
  add it through the debian packaging.

KdeSudo 3.3.1
-----------
* POT file added for translation

KdeSudo 3.3
-----------
* Fixes LP bugs #272427, #268092, #267311, #267501

KdeSudo 3.2
-----------
* Merged changes with 2.x branch

KdeSudo 3.1
-----------
* Fixed icon/name checking when starting an application by its full
  path
* Fixed icon/name checking when using X-KDE-SubstituteUID
* Deals correctly with antislashed quotes:
  - Closes Launchpad #187053

KdeSudo 3.0
-----------
* Initial port to kde4

KdeSudo 2.3
-----------
* Fixed -u option working randomly, due to missing pclose
  call when creating the Xauthority file.

KdeSudo 2.2
-----------
* A little issue with nonewdcop option has been fixed,
  due to sudo removing some environment variables, causing
  failure to use current dcopserver.

kdesudo 2.1
-----------
* Args parsing improved, to be compatible with gksu:
  - Support now complicated command with parsed args

* Uses kprocess instead of shell process:
  - kdesudo now supports processes detaching themselves

* Make -t a fake option for kdesu compatibility again:
  - there is no point hiding StdOut...

* Code cleanup

kdesudo 2.0
-----------
* Args parsing mostly rewritten, fixes potential security issues.

* Connects to a PTY andc orrectly records the tty in /var/run/sudo

* "-s" option (aka -k for sudo) now works due to PTY connexion

* Option "-n" option isn't fake option for kdesu cmdline
    compatibility anymore, and uses "nice" to set command priority.

* Added option "-r" for kdesu cmdline compatibility, uses
    "nice" to force process priority to 50/100 (default)

* Added support for -t kdesu cmdline option, for terminal output

* Added support for -f kdesu cmdline option

* Added support for -n kdesu cmdline option, for non storing the password

* Added support for -r and -p kdesu cmdline option, for "nice" support

* KdeSudo closes correctly after 3 bad passwords given

* Fixes kcm modules support

kdesudo 1.4
-----------
* Removed debug code for -c option

kdesudo 1.3
-----------
* Change to using xauth to dynamically generate xauth-keys.
  Frode M. Døving <frode@ubuntu.com>

* Fix vulnerability issue, as kdesudo didn't quote args
  correctly.

kdesudo 1.2
-----------
* Don't use the user's $HOME, potentially corrupts files

kdesudo 1.1
-----------
* Fixed code to work with a mdern version of sudo

kdesudo 1.0
-----------
* Initial public release



kdesudo/kdesudo.cpp
/***************************************************************************
                          kdesudo.cpp  -  the implementation of the
                                          admin granting sudo widget
                             -------------------
    begin                : Sam Feb 15 15:42:12 CET 2003
    copyright            : (C) 2003 by Robert Gruber
                                       <rgruber@users.sourceforge.net>
                           (C) 2007 by Martin Böhm <martin.bohm@kubuntu.org>
                                       Anthony Mercatante <tonio@kubuntu.org>
                                       Canonical Ltd (Jonathan Riddell
                                                      <jriddell@ubuntu.com>)

 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "kdesudo.h"

#include <QtCore/QDataStream>
#include <QtCore/QDir>
#include <QtCore/QFile>
#include <QtCore/QProcess>
#include <QtCore/QString>
#include <QtCore/QStringList>
#include <QtCore/QTemporaryFile>
#include <QtCore/QTextCodec>

#include <KApplication>
#include <KCmdLineArgs>
#include <KDebug>
#include <KLocale>
#include <KMessageBox>
#include <KPasswordDialog>
#include <KPushButton>
#include <KShell>
#include <KStandardDirs>
#include <KWindowSystem>

#include <sys/stat.h>
#include <sys/types.h>

#include <cstdio>
#include <cstdlib>
#include <csignal>

KdeSudo::KdeSudo(const QString &icon, const QString &appname) :
    QObject(),
    m_process(0),
    m_error(false),
    m_pCookie(new KDESu::KDESuPrivate::KCookie)
{
    KCmdLineArgs *args = KCmdLineArgs::parsedArgs();

    keepPwd = (!args->isSet("n"));
    emptyPwd = args->isSet("s");

    bool realtime = args->isSet("r");
    bool priority = args->isSet("p");
    bool showCommand = (!args->isSet("d"));
    bool changeUID = true;
    bool noExec = false;
    QString runas = args->getOption("u");
    QString cmd;
    int winid = -1;
    bool attach = args->isSet("attach");

    m_dialog = new KPasswordDialog;
    m_dialog->setDefaultButton(KDialog::Ok);

    if (attach) {
        winid = args->getOption("attach").toInt(&attach, 0);
        KWindowSystem::setMainWindow(m_dialog, (WId)winid);
    }

    if (!args->isSet("c") && !args->count() && (!args->isSet("s"))) {
        KMessageBox::information(0, i18n("No command arguments supplied!\n"
                                         "Usage: kdesudo [-u <runas>] <command>\n"
                                         "KdeSudo will now exit...")
                                );
        noExec = true;
    }

    m_process = new KProcess;
    m_process->clearProgram();

    /* load the icon */
    m_dialog->setPixmap(icon);

    // Parsins args

    /* Get the comment out of cli args */
    QByteArray commentBytes = args->getOption("comment").toUtf8();
    QTextCodec *tCodecConv = QTextCodec::codecForLocale();
    QString comment = tCodecConv->toUnicode(commentBytes, commentBytes.size());

    if (args->isSet("f")) {
        // If file is writeable, do not change uid
        QString file = args->getOption("f");
        if (!file.isEmpty()) {
            if (file.at(0) != '/') {
                KStandardDirs dirs;
                file = dirs.findResource("config", file);
                if (file.isEmpty()) {
                    kWarning(1206) << "Config file not found: " << file << "\n";
                    exit(1);
                }
            }
            QFileInfo fi(file);
            if (!fi.exists()) {
                kWarning(1206) << "File does not exist: " << file << "\n";
                exit(1);
            }
            if (fi.isWritable()) {
                changeUID = false;
            }
        }
    }

    connect(m_process, SIGNAL(readyReadStandardOutput()),
            this, SLOT(parseOutput()));

    connect(m_process, SIGNAL(readyReadStandardError()),
            this, SLOT(parseOutput()));

    connect(m_process, SIGNAL(finished(int)),
            this, SLOT(procExited(int)));

    connect(m_dialog, SIGNAL(gotPassword(const QString & , bool)),
            this, SLOT(pushPassword(const QString &)));

    connect(m_dialog, SIGNAL(rejected()),
            this, SLOT(slotCancel()));

    // Generate the xauth cookie and put it in a tempfile
    // set the environment variables to reflect that.
    // Default cookie-timeout is 60 sec. .
    // 'man xauth' for more info on xauth cookies.

    QTemporaryFile *tmpFile = new QTemporaryFile("/tmp/kdesudo-XXXXXX-xauth");
    tmpFile->open();
    QString m_tmpName = tmpFile->fileName();
    delete tmpFile;

    QByteArray disp = m_pCookie->display();

    // Create two processes, one for each xauth call
    QProcess xauth_ext;
    QProcess xauth_merge;

    // This makes "xauth extract - $DISPLAY | xauth -f /tmp/kdesudo-... merge -"
    xauth_ext.setStandardOutputProcess(&xauth_merge);

    // Start the first
    xauth_ext.start("xauth", QStringList() << "extract" << "-" << QString::fromLocal8Bit(disp), QIODevice::ReadOnly);
    if (!xauth_ext.waitForStarted()) {
        return;
    }

    // Start the second
    xauth_merge.start("xauth", QStringList() << "-f" << m_tmpName << "merge" << "-", QIODevice::WriteOnly);
    if (!xauth_merge.waitForStarted()) {
        return;
    }

    // If they ended, close it all
    if (!xauth_merge.waitForFinished()) {
        return;
    }
    xauth_merge.close();

    if (!xauth_ext.waitForFinished()) {
        return;
    }
    xauth_ext.close();

    // non root users need to be able to read the xauth file.
    // the xauth file is deleted when kdesudo exits. security?
    QFile tf;
    tf.setFileName(m_tmpName);

    if (!runas.isEmpty() && runas != "root" && tf.exists()) {
        chmod(QFile::encodeName(m_tmpName), 0644);
    }

    m_process->setEnv("DISPLAY", disp);
    m_process->setEnv("XAUTHORITY", m_tmpName);

    if (emptyPwd) {
        *m_process << "sudo" << "-k";
    } else {
        if (changeUID) {
            *m_process << "sudo" << "-H" << "-S" << "-p" << "passprompt";

            if (!runas.isEmpty()) {
                *m_process << "-u" << runas;
            }
            *m_process << "--";
        }

        if (realtime) {
            *m_process << "nice" << "-n" << "10";
            m_dialog->addCommentLine(i18n("Priority:"), i18n("realtime:") +
                                     QChar(' ') + QString("50/100"));
            *m_process << "--";
        } else if (priority) {
            QString n = args->getOption("p");
            int intn = atoi(n.toUtf8());
            intn = (intn * 40 / 100) - (20 + 0.5);

            QString strn;
            strn.sprintf("%d", intn);

            *m_process << "nice" << "-n" << strn;
            m_dialog->addCommentLine(i18n("Priority:"), n + QString("/100"));
            *m_process << "--";
        }



        if (args->isSet("c")) {
            QString command = args->getOption("c");
            cmd += command;
            *m_process << "sh";
            *m_process << "-c";
            *m_process << command;
        }

        else if (args->count()) {
            for (int i = 0; i < args->count(); i++) {
                if ((!args->isSet("c")) && (i == 0)) {
                    QStringList argsSplit = KShell::splitArgs(args->arg(i));
                    for (int j = 0; j < argsSplit.count(); j++) {
                        *m_process << validArg(argsSplit[j]);
                        if (j == 0) {
                            cmd += validArg(argsSplit[j]) + QChar(' ');
                        } else {
                            cmd += KShell::quoteArg(validArg(argsSplit[j])) + QChar(' ');
                        }
                    }
                } else {
                    *m_process << validArg(args->arg(i));
                    cmd += validArg(args->arg(i)) + QChar(' ');
                }
            }
        }
        // strcmd needs to be defined
        if (showCommand && !cmd.isEmpty()) {
            m_dialog->addCommentLine(i18n("Command:"), cmd);
        }
    }

    if (comment.isEmpty()) {
        QString defaultComment = "<b>%1</b> " + i18n("needs administrative privileges. ");

        if (runas.isEmpty() || runas == "root") {
            defaultComment += i18n("Please enter your password.");
        } else {
            defaultComment += i18n("Please enter password for <b>%1</b>.", runas);
        }

        if (!appname.isEmpty()) {
            m_dialog->setPrompt(defaultComment.arg(appname));
        } else {
            m_dialog->setPrompt(defaultComment.arg(cmd));
        }
    } else {
        m_dialog->setPrompt(comment);
    }

    m_process->setOutputChannelMode(KProcess::MergedChannels);

    if (noExec) {
        exit(0);
    } else {
        m_process->start();
    }
}

KdeSudo::~KdeSudo()
{
    delete m_dialog;
}

void KdeSudo::error(const QString &msg)
{
    m_error = true;
    KMessageBox::error(0, msg);
    KApplication::kApplication()->exit(1);
}

void KdeSudo::parseOutput()
{
    QString strOut = m_process->readAllStandardOutput();

    static int badpass = 0;

    if (strOut.contains("try again")) {
        badpass++;
        if (badpass == 1) {
            m_dialog->addCommentLine(i18n("<b>Warning: </b>"), i18n("<b>Incorrect password, please try again.</b>"));
            m_dialog->show();
        } else if (badpass == 2) {
            m_dialog->show();
        } else {
            error(i18n("Wrong password! Exiting..."));
        }

    } else if (strOut.contains("command not found")) {
        error(i18n("Command not found!"));
    } else if (strOut.contains("is not in the sudoers file")) {
        error(i18n("Your username is unknown to sudo!"));
    } else if (strOut.contains("is not allowed to execute")) {
        error(i18n("Your user is not allowed to run the specified command!"));
    } else if (strOut.contains("is not allowed to run sudo on")) {
        error(i18n("Your user is not allowed to run sudo on this host!"));
    } else if (strOut.contains("may not run sudo on")) {
        error(i18n("Your user is not allowed to run sudo on this host!"));
    } else if ((strOut.contains("passprompt")) || (strOut.contains("PIN (CHV2)"))) {
        m_dialog->setPassword(QString());
        m_dialog->show();
    } else {
        fprintf(stdout, "%s", strOut.toLocal8Bit().constData());
    }
}

void KdeSudo::procExited(int exitCode)
{
    if (!keepPwd && unCleaned) {
        unCleaned = false;
        m_process->clearProgram(); //clearArguments()
        *m_process << "sudo" << "-k";
        m_process->start();
    }

    if (!m_error) {
        if (!m_tmpName.isEmpty()) {
            QFile::remove(m_tmpName);
        }
    }
    KApplication::kApplication()->exit(exitCode);
}

void KdeSudo::pushPassword(const QString &pwd)
{
    m_process->write(pwd.toLocal8Bit() + "\n");
}

void KdeSudo::slotCancel()
{
    KApplication::kApplication()->exit(1);
}

void KdeSudo::slotUser1()
{
    m_dialog->done(AsUser);
}

void KdeSudo::blockSigChild()
{
    sigset_t sset;
    sigemptyset(&sset);
    sigaddset(&sset, SIGCHLD);
    sigprocmask(SIG_BLOCK, &sset, 0L);
}

void KdeSudo::unblockSigChild()
{
    sigset_t sset;
    sigemptyset(&sset);
    sigaddset(&sset, SIGCHLD);
    sigprocmask(SIG_UNBLOCK, &sset, 0L);
}


QString KdeSudo::validArg(QString arg)
{
    QChar firstChar = arg.at(0);
    QChar lastChar = arg.at(arg.length() - 1);

    if ((firstChar == '"' && lastChar == '"') || (firstChar == '\'' && lastChar == '\'')) {
        arg = arg.remove(0, 1);
        arg = arg.remove(arg.length() - 1, 1);
    }
    return arg;
}
